File: src/adapters/qwencode/install_qwencode_integration.py
"""
QwenCode CLI Inheritance Integration Installation Script
Install cross-CLI collaboration awareness capabilities for QwenCode CLI

Usage:
python install_qwencode_integration.py [--verify|--uninstall]
"""

import os
import sys
import json
import shutil
import argparse
from pathlib import Path
from datetime import datetime

# Get current file directory
current_dir = Path(__file__).parent
project_root = current_dir.parent.parent.parent

# QwenCode CLI configuration paths
QWENCODE_CONFIG_DIR = os.path.expanduser("~/.config/qwencode")
QWENCODE_CONFIG_FILE = os.path.join(QWENCODE_CONFIG_DIR, "config.yml")

def create_qwencode_config_directory():
    """Create QwenCode configuration directory"""
    os.makedirs(QWENCODE_CONFIG_DIR, exist_ok=True)
    print(f"[OK] Created QwenCode configuration directory: {QWENCODE_CONFIG_DIR}")

def install_qwencode_plugins():
    """Install QwenCode Plugin configuration"""
    # Read existing config configuration
    existing_config = {}
    if os.path.exists(QWENCODE_CONFIG_FILE):
        try:
            import yaml
            with open(QWENCODE_CONFIG_FILE, 'r', encoding='utf-8') as f:
                existing_config = yaml.safe_load(f) or {}
        except Exception as e:
            print(f"⚠️ Failed to read existing config configuration: {e}")
            existing_config = {}

    # Define cross-CLI collaboration Plugin configuration
    cross_cli_plugins = {
        "cross_cli_inheritance_adapter": {
            "name": "CrossCLIAdapterPlugin",
            "module": "src.adapters.qwencode.inheritance_adapter",
            "class": "QwenCodeInheritanceAdapter",
            "enabled": True,
            "priority": 100,
            "base_class": "BaseQwenCodePlugin",
            "handlers": [
                "on_prompt_received",
                "on_code_generated",
                "on_error_occurred",
                "on_file_created",
                "on_before_save"
            ],
            "config": {
                "cross_cli_enabled": True,
                "supported_clis": ["claude", "gemini", "iflow", "qoder", "codebuddy", "copilot"],
                "auto_detect": True,
                "timeout": 30,
                "error_handling": "continue",
                "collaboration_mode": "active"
            }
        }
    }

    # Merge configuration (preserve existing configuration, add collaboration features)
    merged_config = existing_config.copy()
    if 'plugins' not in merged_config:
        merged_config['plugins'] = []

    # Check if cross-CLI plugin already exists
    existing_plugins = merged_config.get('plugins', [])
    cross_cli_plugin_exists = any(
        plugin.get('name') == 'CrossCLIAdapterPlugin'
        for plugin in existing_plugins
    )

    if not cross_cli_plugin_exists:
        merged_config['plugins'].append(cross_cli_plugins['cross_cli_inheritance_adapter'])

    # Write config configuration file
    try:
        import yaml
        with open(QWENCODE_CONFIG_FILE, 'w', encoding='utf-8') as f:
            yaml.dump(merged_config, f, default_flow_style=False, allow_unicode=True)

        print(f"[OK] QwenCode configuration installed: {QWENCODE_CONFIG_FILE}")
        print("🔗 Installed Plugins:")
        for plugin in merged_config.get('plugins', []):
            if plugin.get('name') == 'CrossCLIAdapterPlugin':
                print(f"   - {plugin['name']}: [OK] Cross-CLI collaboration awareness")

        return True
    except Exception as e:
        print(f"[ERROR] Failed to install QwenCode configuration: {e}")
        return False

def copy_adapter_file():
    """Copy adapter files to QwenCode configuration directory"""
    try:
        # 创建适配器目录
        adapter_dir = os.path.join(QWENCODE_CONFIG_DIR, "plugins")
        os.makedirs(adapter_dir, exist_ok=True)

        # Copy adapter files
        adapter_files = [
            "inheritance_adapter.py",
            "standalone_qwencode_adapter.py"
        ]

        for file_name in adapter_files:
            src_file = current_dir / file_name
            dst_file = os.path.join(adapter_dir, file_name)

            if src_file.exists():
                shutil.copy2(src_file, dst_file)
                print(f"[OK] Copied adapter file: {file_name}")
            else:
                print(f"[WARNING] Adapter file does not exist: {file_name}")

        return True
    except Exception as e:
        print(f"[ERROR] Failed to copy adapter files: {e}")
        return False

def verify_installation():
    """Verify installation was successful"""
    print("\n[VERIFY] Verifying QwenCode CLI integration installation...")

    # Check configuration directory
    if not os.path.exists(QWENCODE_CONFIG_DIR):
        print(f"[ERROR] Configuration directory does not exist: {QWENCODE_CONFIG_DIR}")
        return False

    # Check configuration file
    if not os.path.exists(QWENCODE_CONFIG_FILE):
        print(f"[ERROR] Configuration file does not exist: {QWENCODE_CONFIG_FILE}")
        return False

    # Check adapter directory
    adapter_dir = os.path.join(QWENCODE_CONFIG_DIR, "plugins")
    if not os.path.exists(adapter_dir):
        print(f"[ERROR] Adapter directory does not exist: {adapter_dir}")
        return False

    # Read and verify configuration
    try:
        import yaml
        with open(QWENCODE_CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = yaml.safe_load(f) or {}

        plugins = config.get('plugins', [])
        cross_cli_plugin = None

        for plugin in plugins:
            if plugin.get('name') == 'CrossCLIAdapterPlugin':
                cross_cli_plugin = plugin
                break

        if cross_cli_plugin:
            plugin_config = cross_cli_plugin.get('config', {})
            if plugin_config.get('cross_cli_enabled'):
                print("[SUCCESS] Cross-CLI collaboration plugin: Enabled")
                print("[SUCCESS] Supported CLI tools:")
                supported_clis = plugin_config.get('supported_clis', [])
                for cli in supported_clis:
                    print(f"   - {cli}")
                print("[SUCCESS] Auto-detection: Enabled")
                return True
            else:
                print("[WARNING] Cross-CLI collaboration plugin: Disabled")
                return False
        else:
            print("[ERROR] Cross-CLI collaboration plugin: Not found")
            return False

    except Exception as e:
        print(f"[ERROR] Failed to verify configuration: {e}")
        return False

def uninstall_qwencode_integration():
    """Uninstall QwenCode integration"""
    try:
        # Backup existing configuration
        if os.path.exists(QWENCODE_CONFIG_FILE):
            backup_file = f"{QWENCODE_CONFIG_FILE}.backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
            shutil.copy2(QWENCODE_CONFIG_FILE, backup_file)
            print(f"[BACKUP] Backed up existing configuration: {backup_file}")

        # Remove adapter directory
        adapter_dir = os.path.join(QWENCODE_CONFIG_DIR, "plugins")
        if os.path.exists(adapter_dir):
            shutil.rmtree(adapter_dir)
            print(f"[DELETE] Deleted adapter directory: {adapter_dir}")

        print("[OK] QwenCode integration uninstalled")
        return True
    except Exception as e:
        print(f"[ERROR] Uninstall failed: {e}")
        return False

def main():
    parser = argparse.ArgumentParser(
        description="QwenCode CLI Cross-CLI Collaboration Integration Installation Script",
        formatter_class=argparse.RawDescriptionHelpFormatter
    )

    parser.add_argument(
        "--install",
        action="store_true",
        help="Install QwenCode CLI Cross-CLI Collaboration Integration"
    )

    parser.add_argument(
        "--verify",
        action="store_true",
        help="Verify QwenCode CLI integration installation"
    )

    parser.add_argument(
        "--uninstall",
        action="store_true",
        help="Uninstall QwenCode CLI Cross-CLI Collaboration Integration"
    )

    args = parser.parse_args()

    print("[INSTALL] QwenCode CLI Cross-CLI Collaboration Integration Installer")
    print("=" * 60)

    if args.uninstall:
        print("[UNINSTALL] Uninstall mode...")
        success = uninstall_qwencode_integration()
    elif args.verify:
        print("[VERIFY] Verification mode...")
        success = verify_installation()
    elif args.install or len(sys.argv) == 1:
        print("[INSTALL] Installation mode...")

        # 1. Create configuration directory
        print("Step 1. Create configuration directory...")
        create_qwencode_config_directory()

        # 2. Install plugin configuration
        print("Step 2. Install plugin configuration...")
        config_success = install_qwencode_plugins()

        # 3. Copy adapter files
        print("Step 3. Copy adapter files...")
        adapter_success = copy_adapter_file()

        success = config_success and adapter_success

        if success:
            print("\n[SUCCESS] QwenCode CLI integration installed successfully!")
            print("\n[INFO] Installation Summary:")
            print(f"   [SUCCESS] 配置目录: {QWENCODE_CONFIG_DIR}")
            print(f"   [SUCCESS] 配置文件: {QWENCODE_CONFIG_FILE}")
            print(f"   [SUCCESS] 适配器目录: {os.path.join(QWENCODE_CONFIG_DIR, 'plugins')}")
            print("   [SUCCESS] Cross-CLI collaboration: Enabled")

            print("\n[INFO] Next steps:")
            print("   1. 安装其他CLI工具的集成: ai-cli-router deploy --all")
            print("   2. 初始化项目: ai-cli-router init")
            print("   3. 开始协作: qwencode-cli '请用gemini帮我分析代码'")
        else:
            print("\n[ERROR] QwenCode CLI integration installation failed, please check error messages")
    else:
        parser.print_help()

if __name__ == "__main__":
    main()

File: src/adapters/qwencode/install_qwencode_integration.js
#!/usr/bin/env node

/**
 * QwenCode CLI Inheritance Integration Installation Script
 * Install cross-CLI collaboration awareness capabilities for QwenCode CLI
 * 
 * Usage:
 * node install_qwencode_integration.js [--verify|--uninstall]
 */

import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import { homedir } from 'os';
import { execSync } from 'child_process';

// Get current file directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.join(__dirname, '..', '..', '..');

// QwenCode CLI configuration paths
const QWENCODE_CONFIG_DIR = path.join(homedir(), '.config', 'qwencode');
const QWENCODE_CONFIG_FILE = path.join(QWENCODE_CONFIG_DIR, 'config.yml');

// Check if yaml library is installed
let yaml;
try {
    yaml = await import('js-yaml');
} catch (error) {
    console.warn('⚠️ js-yaml library not found, attempting to install...');
    try {
        execSync('npm install js-yaml', { stdio: 'inherit' });
        yaml = await import('js-yaml');
        console.log('[OK] js-yaml library installed successfully');
    } catch (installError) {
        console.error('❌ Unable to install js-yaml library, please install manually: npm install js-yaml');
        process.exit(1);
    }
}

async function createQwenCodeConfigDirectory() {
    /** Create QwenCode configuration directory */
    try {
        await fs.mkdir(QWENCODE_CONFIG_DIR, { recursive: true });
        console.log(`[OK] Created QwenCode configuration directory: ${QWENCODE_CONFIG_DIR}`);
    } catch (error) {
        console.error(`[ERROR] Failed to create QwenCode configuration directory: ${error.message}`);
    }
}

async function installQwenCodePlugins() {
    /** Install QwenCode Plugin configuration */
    // Read existing config configuration
    let existingConfig = {};
    try {
        const configExists = await fs.access(QWENCODE_CONFIG_FILE).then(() => true).catch(() => false);
        if (configExists) {
            const configContent = await fs.readFile(QWENCODE_CONFIG_FILE, 'utf8');
            existingConfig = yaml.load(configContent) || {};
        }
    } catch (error) {
        console.warn(`⚠️ Failed to read existing config configuration: ${error.message}`);
        existingConfig = {};
    }

    // Define cross-CLI collaboration Plugin configuration
    const crossCliPlugins = {
        "cross_cli_inheritance_adapter": {
            "name": "CrossCLIAdapterPlugin",
            "module": "src.adapters.qwencode.inheritance_adapter",
            "class": "QwenCodeInheritanceAdapter",
            "enabled": true,
            "priority": 100,
            "base_class": "BaseQwenCodePlugin",
            "handlers": [
                "on_prompt_received",
                "on_code_generated",
                "on_error_occurred",
                "on_file_created",
                "on_before_save"
            ],
            "config": {
                "cross_cli_enabled": true,
                "supported_clis": ["claude", "gemini", "iflow", "qoder", "codebuddy", "copilot"],
                "auto_detect": true,
                "timeout": 30,
                "error_handling": "continue",
                "collaboration_mode": "active"
            }
        }
    };

    // Merge configuration (preserve existing configuration, add collaboration features)
    const mergedConfig = { ...existingConfig };
    if (!mergedConfig.plugins) {
        mergedConfig.plugins = [];
    }

    // Check if cross-CLI plugin already exists
    const existingPlugins = mergedConfig.plugins || [];
    const crossCliPluginExists = existingPlugins.some(
        plugin => plugin.name === 'CrossCLIAdapterPlugin'
    );

    if (!crossCliPluginExists) {
        mergedConfig.plugins.push(crossCliPlugins.cross_cli_inheritance_adapter);
    }

    // Write config configuration file
    try {
        const yamlContent = yaml.dump(mergedConfig, {
            lineWidth: -1,
            noRefs: true,
            quotingType: '"'
        });
        
        await fs.writeFile(QWENCODE_CONFIG_FILE, yamlContent, 'utf8');
        console.log(`[OK] QwenCode configuration installed: ${QWENCODE_CONFIG_FILE}`);
        console.log("🔗 Installed Plugins:");
        
        for (const plugin of mergedConfig.plugins) {
            if (plugin.name === 'CrossCLIAdapterPlugin') {
                console.log(`   - ${plugin.name}: [OK] Cross-CLI collaboration awareness`);
            }
        }
        
        return true;
    } catch (error) {
        console.error(`❌ Failed to install QwenCode configuration: ${error.message}`);
        return false;
    }
}

async function copyAdapterFiles() {
    /** Copy adapter files to QwenCode configuration directory */
    try {
        // Create adapter directory
        const adapterDir = path.join(QWENCODE_CONFIG_DIR, 'adapters');
        await fs.mkdir(adapterDir, { recursive: true });

        // Copy adapter files
        const adapterFiles = [
            'inheritance_adapter.py'
        ];

        for (const fileName of adapterFiles) {
            const srcFile = path.join(__dirname, fileName);
            const dstFile = path.join(adapterDir, fileName);

            try {
                await fs.access(srcFile);
                await fs.copyFile(srcFile, dstFile);
                console.log(`[OK] Copied adapter file: ${fileName}`);
            } catch (error) {
                console.warn(`⚠️ Adapter file does not exist: ${fileName}`);
            }
        }

        return true;
    } catch (error) {
        console.error(`❌ Failed to copy adapter files: ${error.message}`);
        return false;
    }
}

async function verifyInstallation() {
    /** Verify if installation was successful */
    console.log('\n🔍 Verifying QwenCode CLI integration installation...');

    // Check configuration directory
    try {
        await fs.access(QWENCODE_CONFIG_DIR);
    } catch (error) {
        console.error(`❌ Configuration directory does not exist: ${QWENCODE_CONFIG_DIR}`);
        return false;
    }

    // Check config file
    try {
        await fs.access(QWENCODE_CONFIG_FILE);
    } catch (error) {
        console.error(`❌ Config configuration file does not exist: ${QWENCODE_CONFIG_FILE}`);
        return false;
    }

    // Check adapter directory
    const adapterDir = path.join(QWENCODE_CONFIG_DIR, 'adapters');
    try {
        await fs.access(adapterDir);
    } catch (error) {
        console.error(`❌ Adapter directory does not exist: ${adapterDir}`);
        return false;
    }

    // Read and verify config configuration
    try {
        const configContent = await fs.readFile(QWENCODE_CONFIG_FILE, 'utf8');
        const config = yaml.load(configContent);

        // Check if key plugin exists
        const plugins = config.plugins || [];
        const hasCrossCliPlugin = plugins.some(plugin => plugin.name === 'CrossCLIAdapterPlugin');
        
        if (!hasCrossCliPlugin) {
            console.warn('⚠️ Missing cross-CLI collaboration plugin: CrossCLIAdapterPlugin');
        }

        console.log('[OK] QwenCode CLI integration installation verification passed');
        return true;
    } catch (error) {
        console.error(`❌ Failed to verify config configuration: ${error.message}`);
        return false;
    }
}

async function uninstallQwenCodeIntegration() {
    /** Uninstall QwenCode integration */
    try {
        // Delete config configuration
        try {
            await fs.unlink(QWENCODE_CONFIG_FILE);
            console.log(`[OK] Deleted QwenCode Config configuration: ${QWENCODE_CONFIG_FILE}`);
        } catch (error) {
            if (error.code !== 'ENOENT') {
                console.warn(`⚠️ Failed to delete Config configuration: ${error.message}`);
            }
        }

        // Delete adapter directory
        const adapterDir = path.join(QWENCODE_CONFIG_DIR, 'adapters');
        try {
            await fs.rm(adapterDir, { recursive: true, force: true });
            console.log(`[OK] Deleted QwenCode adapter directory: ${adapterDir}`);
        } catch (error) {
            if (error.code !== 'ENOENT') {
                console.warn(`⚠️ Failed to delete adapter directory: ${error.message}`);
            }
        }

        console.log('[OK] QwenCode CLI integration uninstallation completed');
        return true;
    } catch (error) {
        console.error(`❌ Failed to uninstall QwenCode integration: ${error.message}`);
        return false;
    }
}

async function main() {
    /** Main function */
    const args = process.argv.slice(2);
    const options = {
        verify: args.includes('--verify'),
        uninstall: args.includes('--uninstall'),
        install: args.includes('--install') || args.length === 0
    };

    console.log('QwenCode CLI Cross-CLI Collaboration Integration Installer');
    console.log('='.repeat(50));

    if (options.uninstall) {
        console.log('[UNINSTALL] Uninstall mode...');
        await uninstallQwenCodeIntegration();
    } else if (options.verify) {
        console.log('🔍 Verification mode...');
        await verifyInstallation();
    } else if (options.install) {
        console.log('📦 Installation mode...');
        
        // 1. Create configuration directory
        await createQwenCodeConfigDirectory();

        // 2. Install Plugin configuration
        const pluginSuccess = await installQwenCodePlugins();

        // 3. Copy adapter files
        const adapterSuccess = await copyAdapterFiles();

        const success = pluginSuccess && adapterSuccess;

        if (success) {
            console.log('\n🎉 QwenCode CLI Cross-CLI Collaboration Integration Installation Successful!');
            console.log('\n[INFO] Installation Summary:');
            console.log(`   [OK] Configuration Directory: ${QWENCODE_CONFIG_DIR}`);
            console.log(`   [OK] Config File: ${QWENCODE_CONFIG_FILE}`);
            console.log(`   [OK] Adapter Directory: ${path.join(QWENCODE_CONFIG_DIR, 'adapters')}`);
            console.log(`   [OK] Cross-CLI Collaboration Plugin: Enabled`);
            
            console.log('\n[INSTALL] Next Steps:');
            console.log('   1. Run installation scripts for other CLI tools');
            console.log('   2. Use stigmergy-cli deploy --all to install all tools');
            console.log('   3. Use stigmergy-cli init to initialize the project');
        } else {
            console.log('\n❌ QwenCode CLI Cross-CLI Collaboration Integration Installation Failed');
        }
    } else {
        console.log('Usage:');
        console.log('  node install_qwencode_integration.js [--install|--verify|--uninstall]');
        console.log('  Default is installation mode');
    }
}

// Run main function
if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(error => {
        console.error(`[FATAL] ${error.message}`);
        process.exit(1);
    });
}

export { 
    createQwenCodeConfigDirectory, 
    installQwenCodePlugins, 
    copyAdapterFiles, 
    verifyInstallation, 
    uninstallQwenCodeIntegration 
};